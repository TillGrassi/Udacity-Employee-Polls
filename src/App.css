.App {
  text-align: center;
}

/* LogIn */
.LogIn {
  display: flex;
  flex-direction: column;
  text-align: center;
  align-items: center;
}

.LoginForm {
  width: 50vh;
  height: 25px;
  margin: 10px;
}

/* NavBar */
.Navbar {
  display: block;
  list-style-type: none;
  height: 40px;
  margin: 0;
  padding: 0;
}

.NavLink {
  display: block;
  padding: 8px;
}

.NavLeft {
  float: left;
}

.NavRight {
  float: right;
}

.NavAvatar {
  margin: 1px;
  margin-top: 5px;
  height: 28px;
  width: 28px;
}

/*Dashboard*/

.PollsContainer {
  margin: 5px 30px 5px 30px;
  border: 2px solid black;
}

h4 {
  padding: 5px 0px 5px 0px;
  margin: 0;
  border-bottom: 2px solid black;
}

.Poll {
  height: 100px;
  width: 150px;
  margin: 5px;
  padding: 2px;
  border: 2px solid black;
}

.PollsFlex {
  display: flex;
}

p {
  margin: 5px;
}