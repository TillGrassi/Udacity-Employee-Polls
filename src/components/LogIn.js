import { connect } from "react-redux";
import { useState } from "react";
import { setAuthedUser } from "../actions/authedUser";
import { useNavigate } from "react-router-dom";
import sarahedo from "../avatar/avatar1.jpg";
import tylermcginnis from "../avatar/avatar2.jpg";
import mtsamis from "../avatar/avatar3.jpg";
import zoshikanlu from "../avatar/avatar4.jpg";

const Login = ({ users, dispatch }) => {
  const [userPicked, setUserPicked] = useState("none");
  const navigate = useNavigate();
  const handleLogIn = (e) => {
    e.preventDefault();

    dispatch(setAuthedUser(userPicked));
    navigate("/");
  };

  return (
    <form className="LogIn">
      <h1>Select the user:</h1>
      {userPicked === "sarahedo" && <img src={sarahedo} />}
      {userPicked === "tylermcginnis" && <img src={tylermcginnis} />}
      {userPicked === "mtsamis" && <img src={mtsamis} />}
      {userPicked === "zoshikanlu" && <img src={zoshikanlu} />}
      <select
        data-testid="selectUser"
        className="LoginForm"
        onChange={(e) => setUserPicked(e.target.value)}
        value={userPicked}
      >
        <option value="none" disabled>
          Select...
        </option>
        {users.map((user) => (
          <option key={user.id} value={user.id}>
            {user.name}
          </option>
        ))}
      </select>
      <input
        data-testid="password"
        type="password"
        placeholder="password"
        className="LoginForm"
        required
      />
      {userPicked === "none" ? (
        <button disabled className="LoginForm">
          Submit
        </button>
      ) : (
        <button onClick={handleLogIn} className="LoginForm">
          Submit
        </button>
      )}
    </form>
  );
};

const mapStateToProps = ({ users }) => {
  return {
    users: Object.keys(users).map(function (key) {
      return users[key];
    }),
  };
};

export default connect(mapStateToProps)(Login);
