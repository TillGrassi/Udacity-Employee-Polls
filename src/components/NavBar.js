import { connect, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { removeAuthedUser } from "../actions/authedUser";
import sarahedo from "../avatar/avatar1.jpg";
import tylermcginnis from "../avatar/avatar2.jpg";
import mtsamis from "../avatar/avatar3.jpg";
import zoshikanlu from "../avatar/avatar4.jpg";

function NavBar({ authedUser }) {
  const dispatch = useDispatch();
  const handleLogout = () => {
    dispatch(removeAuthedUser());
  };
  return (
    <ul className="Navbar">
      <li className="NavLeft">
        <Link to="/" className="NavLink Link">
          Home
        </Link>
      </li>
      <li className="NavLeft">
        <Link to="/Leaderboard" className="NavLink Link">
          LeaderBoard
        </Link>
      </li>
      <li className="NavLeft">
        <Link to="/NewPoll" className="NavLink Link">
          New
        </Link>
      </li>
      <li className="NavRight">
        <a onClick={handleLogout} className="NavLink">
          Logout
        </a>
      </li>
      <li className="NavRight NavLink">{authedUser.id}</li>
      <li className="NavRight">
        {authedUser.id === "sarahedo" && (
          <img src={sarahedo} className="NavAvatar" />
        )}
        {authedUser.id === "tylermcginnis" && (
          <img src={tylermcginnis} className="NavAvatar" />
        )}
        {authedUser.id === "mtsamis" && (
          <img src={mtsamis} className="NavAvatar" />
        )}
        {authedUser.id === "zoshikanlu" && (
          <img src={zoshikanlu} className="NavAvatar" />
        )}
      </li>
    </ul>
  );
}

const mapStateToProps = ({ authedUser }) => {
  return {
    authedUser,
  };
};

export default connect(mapStateToProps)(NavBar);
