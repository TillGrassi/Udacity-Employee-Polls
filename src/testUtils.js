import { createStore } from "redux";
import middleware from "./middleware";
import reducer from "./reducers";

export function createTestStore() {
  const store = createStore(reducer, middleware);
  return store;
}
export function createLoggedInTestStore() {
  const store = createStore(
    reducer,
    { authedUser: { id: "sarahedo" } },
    middleware
  );
  return store;
}
