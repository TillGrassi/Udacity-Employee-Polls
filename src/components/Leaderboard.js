import { useEffect, useState } from "react";
import { connect } from "react-redux";
import NavBar from "./NavBar";
import Table from "react-bootstrap/table";
import sarahedo from "../avatar/avatar1.jpg";
import tylermcginnis from "../avatar/avatar2.jpg";
import mtsamis from "../avatar/avatar3.jpg";
import zoshikanlu from "../avatar/avatar4.jpg";

function Leaderboard({ users, authedUser }) {
  const [getRanking, setGetRanking] = useState([]);
  useEffect(() => {
    setGetRanking(
      users
        .map((user) => {
          const answerCount = Object.keys(user.answers);
          return {
            name: user.name,
            answers: answerCount.length,
            questions: user.questions.length,
            combined: answerCount.length + user.questions.length,
          };
        })
        .sort((a, b) => b.combined - a.combined)
    );
  }, [users, authedUser]);

  return (
    <div>
      <NavBar />
      <Table data-testid="leaderboard" className="Table">
        <thead>
          <tr>
            <th>#</th>
            <th>Name</th>
            <th>Questions</th>
            <th>Answers</th>
          </tr>
        </thead>
        <tbody>
          {getRanking.map((user) => {
            return (
              <tr key={getRanking.indexOf(user)}>
                <td>{getRanking.indexOf(user) + 1}</td>
                <td>
                  {user.name === "Sarah Edo" && (
                    <img src={sarahedo} alt="" className="NavAvatar" />
                  )}
                  {user.name === "Mike Tsamis" && (
                    <img src={mtsamis} alt="" className="NavAvatar" />
                  )}
                  {user.name === "Tyler McGinnis" && (
                    <img src={tylermcginnis} alt="" className="NavAvatar" />
                  )}
                  {user.name === "Zenobia Oshikanlu" && (
                    <img src={zoshikanlu} alt="" className="NavAvatar" />
                  )}
                  {user.name}
                </td>
                <td>{user.questions}</td>
                <td>{user.answers}</td>
              </tr>
            );
          })}
        </tbody>
      </Table>
    </div>
  );
}

const mapStateToProps = ({ users, authedUser }) => {
  return {
    users: Object.keys(users).map(function (key) {
      return users[key];
    }),
    authedUser,
  };
};
export default connect(mapStateToProps)(Leaderboard);
