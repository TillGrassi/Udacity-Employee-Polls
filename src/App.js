import "./App.css";
import LogIn from "./components/LogIn";
import Dashboard from "./components/Dashboard";
import Pollpage from "./components/Pollpage";
import Leaderboard from "./components/Leaderboard";
import NewPoll from "./components/NewPoll";
import { useEffect } from "react";
import { connect, useDispatch } from "react-redux";
import { handleInitialData } from "./actions/shared";
import { Routes, Route } from "react-router-dom";
import NotFound from "./components/NotFound";

function App(props) {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(handleInitialData());
  }, [dispatch]);

  return (
    <div className="App">
      {props.authedUser.id === null ? (
        <LogIn />
      ) : (
        <Routes>
          <Route path="/" exact element={<Dashboard />} />
          <Route path="/leaderboard" element={<Leaderboard />} />
          <Route path="/add" element={<NewPoll />} />
          <Route path="/questions/:id" exact element={<Pollpage />} />
          <Route path="*" element={<NotFound />} />
        </Routes>
      )}
    </div>
  );
}

const mapStateToProps = ({ authedUser }) => {
  return {
    authedUser,
  };
};

export default connect(mapStateToProps)(App);
