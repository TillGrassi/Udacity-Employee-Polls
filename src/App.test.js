import * as React from "react";
import Leaderboard from "./components/Leaderboard";
import App from "./App";
import LogIn from "./components/LogIn";
import NavBar from "./components/NavBar";
import { render, fireEvent, cleanup, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { createLoggedInTestStore, createTestStore } from "./testUtils";
import { BrowserRouter } from "react-router-dom";

const { _saveQuestion, _saveQuestionAnswer } = require("./_DATA");
let loggedInStore;
let store;
beforeEach(() => {
  store = createTestStore();
  loggedInStore = createLoggedInTestStore();
  cleanup();
});

describe("saving Question", () => {
  it("will return the new question", async () => {
    const question = {
      optionOneText: "Test One",
      optionTwoText: "Test Two",
      author: "sarahedo",
    };
    const result = await _saveQuestion(question);
    expect(result.author).toEqual("sarahedo");
    expect(result.optionOne.text).toEqual("Test One");
    expect(result.optionTwo.text).toEqual("Test Two");
  });
  it("will throw an error when wrong formatted data are given", async () => {
    const question = {
      optionOneText: "Test One",
      optionTwoText: "Test Two",
    };
    try {
      await _saveQuestion(question);
    } catch (e) {
      (e) =>
        expect(e).toEqual(
          "Please provide optionOneText, optionTwoText, and author"
        );
    }
  });
});

describe("saving Question Answer", () => {
  it("saves the correct data at the correct places", async () => {
    const data = {
      authedUser: "sarahedo",
      qid: "vthrdm985a262al8qx3do",
      answer: "optionOne",
    };
    const result = await _saveQuestionAnswer(data);
    expect(result).toEqual(true);
  });
  it("throws an error when given incorrectly formatted data", async () => {
    const data = {
      authedUser: "sarahedo",
      qid: "vthrdm985a262al8qx3do",
    };
    try {
      await _saveQuestionAnswer(data);
    } catch (e) {
      expect(e).toEqual("Please provide authedUser, qid, and answer");
    }
  });
});

describe("Leaderboard", () => {
  it("will match snapshot", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={store}>
          <Leaderboard />
        </Provider>
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("LogInPage", () => {
  it("will match snapshot", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={store}>
          <LogIn />
        </Provider>
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("login page works as expected", () => {
  it("shows all the input fields on the page", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={store}>
          <LogIn />
        </Provider>
      </BrowserRouter>
    );
    const userSelector = component.getByTestId("selectUser");
    const submitButton = component.getByText("Submit");
    expect(userSelector).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });
  it("renders the login page when logging out", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={loggedInStore}>
          <App />
        </Provider>
      </BrowserRouter>
    );
    const dashboard = component.getByText("New Questions");
    expect(dashboard).toBeInTheDocument();
    const logOut = component.getByText("Logout");
    fireEvent.click(logOut);
    const logInSelector = screen.getByTestId("selectUser");
    expect(logInSelector).toBeInTheDocument();
  });
});

describe("The navbar works as expected", () => {
  it("renders with all the links", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={store}>
          <NavBar />
        </Provider>
      </BrowserRouter>
    );
    const home = component.getByText("Home");
    const leaderboard = component.getByText("LeaderBoard");
    const newPoll = component.getByText("New");
    const logout = component.getByText("Logout");
    expect(home).toBeInTheDocument();
    expect(leaderboard).toBeInTheDocument();
    expect(newPoll).toBeInTheDocument();
    expect(logout).toBeInTheDocument();
  });
  it("renders the leaderboard when clicking on the navbar link", () => {
    const component = render(
      <BrowserRouter>
        <Provider store={loggedInStore}>
          <App />
        </Provider>
      </BrowserRouter>
    );
    const leaderboard = component.getByText("LeaderBoard");
    fireEvent.click(leaderboard);
    const leaderboardcomponent = screen.getByTestId("leaderboard");
    expect(leaderboardcomponent).toBeInTheDocument();
  });
});
