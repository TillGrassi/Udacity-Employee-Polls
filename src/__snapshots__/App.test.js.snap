// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <ul
          class="Navbar"
        >
          <li
            class="NavLeft"
          >
            <a
              class="NavLink Link"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="NavLeft"
          >
            <a
              class="NavLink Link"
              href="/Leaderboard"
            >
              LeaderBoard
            </a>
          </li>
          <li
            class="NavLeft"
          >
            <a
              class="NavLink Link"
              href="/NewPoll"
            >
              New
            </a>
          </li>
          <li
            class="NavRight"
          >
            <a
              class="NavLink"
            >
              Logout
            </a>
          </li>
          <li
            class="NavRight NavLink"
          />
          <li
            class="NavRight"
          />
        </ul>
        <table
          class="Table table"
          data-testid="leaderboard"
        >
          <thead>
            <tr>
              <th>
                #
              </th>
              <th>
                Name
              </th>
              <th>
                Questions
              </th>
              <th>
                Answers
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <ul
        class="Navbar"
      >
        <li
          class="NavLeft"
        >
          <a
            class="NavLink Link"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="NavLeft"
        >
          <a
            class="NavLink Link"
            href="/Leaderboard"
          >
            LeaderBoard
          </a>
        </li>
        <li
          class="NavLeft"
        >
          <a
            class="NavLink Link"
            href="/NewPoll"
          >
            New
          </a>
        </li>
        <li
          class="NavRight"
        >
          <a
            class="NavLink"
          >
            Logout
          </a>
        </li>
        <li
          class="NavRight NavLink"
        />
        <li
          class="NavRight"
        />
      </ul>
      <table
        class="Table table"
        data-testid="leaderboard"
      >
        <thead>
          <tr>
            <th>
              #
            </th>
            <th>
              Name
            </th>
            <th>
              Questions
            </th>
            <th>
              Answers
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LogInPage will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="LogIn"
      >
        <h1>
          Select the user:
        </h1>
        <select
          class="LoginForm"
          data-testid="selectUser"
        >
          <option
            disabled=""
            value="none"
          >
            Select...
          </option>
        </select>
        <input
          class="LoginForm"
          data-testid="password"
          placeholder="password"
          required=""
          type="password"
        />
        <button
          class="LoginForm"
          disabled=""
        >
          Submit
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="LogIn"
    >
      <h1>
        Select the user:
      </h1>
      <select
        class="LoginForm"
        data-testid="selectUser"
      >
        <option
          disabled=""
          value="none"
        >
          Select...
        </option>
      </select>
      <input
        class="LoginForm"
        data-testid="password"
        placeholder="password"
        required=""
        type="password"
      />
      <button
        class="LoginForm"
        disabled=""
      >
        Submit
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
